services/command-router/src/main/java/org/eclipse/hono/commandrouter/impl/kafka/MappingAndDelegatingCommandHandler.java

```diff
@@ -0,0 +1,156 @@
+
+        command.logToSpan(currentSpan);
+        if (command.isValid()) {
+            log.trace("received a valid command record");
+            mapAndDelegateIncomingCommand(commandContext);
+        } else {
+            log.debug("malformed command message [tenant: {}, device: {}]", command.getTenant(), command.getDeviceId());
+            commandContext.reject("malformed command message");
+        }
+    }
+
+    private void mapAndDelegateIncomingCommand(final CommandContext commandContext) {
+        final Command command = commandContext.getCommand();
+
+        // determine last used gateway device id
+        log.trace("determine command target gateway/adapter for [{}]", command);
+        commandTargetMapper
+                .getTargetGatewayAndAdapterInstance(command.getTenant(), command.getDeviceId(), commandContext.getTracingContext())
+                .onSuccess(result -> {
+                    final String targetAdapterInstanceId = result
+                            .getString(DeviceConnectionConstants.FIELD_ADAPTER_INSTANCE_ID);
+                    final String targetDeviceId = result.getString(DeviceConnectionConstants.FIELD_PAYLOAD_DEVICE_ID);
+                    final String targetGatewayId = targetDeviceId.equals(command.getDeviceId()) ? null : targetDeviceId;
+
+                    if (Objects.isNull(targetGatewayId)) {
+                        log.trace("command not mapped to gateway, use original device id [{}]", command.getDeviceId());
+                    } else {
+                        command.setGatewayId(targetGatewayId);
+                        log.trace("determined target gateway [{}] for device [{}]", targetGatewayId,
+                                command.getDeviceId());
+                        commandContext.getTracingSpan().log("determined target gateway [" + targetGatewayId + "]");
+                    }
+
+                    log.trace("delegate command to target adapter instance '{}' [command: {}]", targetAdapterInstanceId,
+                            commandContext.getCommand());
+                    internalCommandSender.sendCommand(commandContext, targetAdapterInstanceId);
+                })
+                .onFailure(cause -> {
+                    final String errorMsg;
+
+                    if (ServiceInvocationException.extractStatusCode(cause) == HttpURLConnection.HTTP_NOT_FOUND) {
+                        errorMsg = "no target adapter instance found for command with device id "
+                                + command.getDeviceId();
+                    } else {
+                        errorMsg = "error getting target gateway and adapter instance for command with device id"
+                                + command.getDeviceId();
+                    }
+                    log.debug(errorMsg, cause);
+                    TracingHelper.logError(commandContext.getTracingSpan(), errorMsg, cause);
+                }).onComplete(ok -> commandContext.release());
```

Comment:
`commandContext.release()` would also be called in the "success" case here which isn't right.
I think this belongs at the end of the `onFailure` block.